### PR TITLE
Fix issue with reuse layers in torch

### PR DESCRIPTION
**Pull request by reuvenperetz**
## Pull Request Description:
Fix the issue of ignoring shared parameters in reused layers in torch models.
The PR also fixes an issue with handling reused nodes during Hessian computation, which replaces reused nodes with their representative 'base' nodes.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 1]
```diff
@@ -249,6 +249,31 @@ def nodes_builder(model: GraphModule,
         # Extract input and output shapes of the node.
         input_shape, output_shape = _extract_input_and_output_shapes(node)
 
+        # Check if this node's target has been seen before
+        reuse = False
+        reuse_group = None
+        node_group_key = create_reuse_group(node.target, weights)
+        # We mark nodes as reused only if there are multiple nodes in the graph with same
+        # 'target' and it has some weights.
+        if node_group_key in seen_targets and len(weights) > 0:
```
Comment (irenaby) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:258]: Is there any reason not to mark all reused nodes, for consistency?

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 2]
```diff
@@ -365,3 +392,25 @@ def edges_builder(model: GraphModule,
                     Edge(fx_node_2_graph_node[node], fx_node_2_graph_node[out_node], src_index, dst_index))
 
     return edges
+
+
+def create_reuse_group(target: Any, weights: Dict[str, Any]) -> str:
+    """
+    Combine target and weights to create a unique reuse group identifier.
```
Comment (irenaby) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:399]: Maybe mention that weights are needed for functional ops

File: tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py [hunk 3]
```diff
@@ -12,56 +12,164 @@
...
+        self.conv1_weight = nn.Parameter(torch.randn(3, 3, 2, 2))
+        self.conv1_bias = nn.Parameter(torch.randn(3))
+        self.conv2_weight = nn.Parameter(torch.randn(3, 3, 2, 2))
+        self.conv2_bias = nn.Parameter(torch.randn(3))
+
+    def forward(self, x):
+        x = F.conv2d(x, self.conv1_weight, self.conv1_bias, stride=1, groups=1)
+        x = F.conv2d(x, self.conv1_weight, self.conv1_bias, stride=1, groups=1)
+        x = F.conv2d(x, self.conv2_weight, self.conv2_bias, stride=1, groups=1)
+        x = F.conv2d(x, self.conv2_weight, self.conv2_bias, stride=1, groups=1)
```
Comment (irenaby) [inline tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py:56]: I would add a case where weight is identical but not bias and vice versa to make sure both are taken into account

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py (hunk 1 above)
Comment (irenaby) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:258]: Is there any reason not to mark all reused layers for consistency?

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py (hunk 2 above)
Comment (irenaby) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:399]: Maybe mention that weights are needed for functional ops

File: tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py (hunk 3 above)
Comment (irenaby) [inline tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py:56]: I would also check that identical weight but different bias and vice versa are not considered reused

File: tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py [hunk 4]
```diff
@@ -12,56 +12,164 @@
...
+
+class ReuseLayerTestBase(BasePytorchTest):
     def __init__(self, unit_test):
         super().__init__(unit_test)
 
     def create_inputs_shape(self):
-        return [[self.val_batch_size, 3, 32, 32], [self.val_batch_size, 3, 32, 32]]
+        return [[self.val_batch_size, 3, 32, 32]]
 
     def create_feature_network(self, input_shape):
```
Comment (irenaby) [inline tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py:81]: Why not implement this method directly instead of making it call another method that should be implemented?

File: tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py [hunk 5]
```diff
@@ -12,56 +12,164 @@
...
-"""
-This test checks:
-The reuse of a layer in a model.
-"""
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
 
-class ReuseLayerNet(torch.nn.Module):
+class BaseReuseLayerNet(nn.Module):
```
Comment (irenaby) [inline tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py:23]: Do you really need this class? I think you are over engineering

File: tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py [hunk 6]
```diff
@@ -12,56 +12,164 @@
...
 
     def create_feature_network(self, input_shape):
-        return ReuseLayerNet()
\ No newline at end of file
+        return self.get_model()
+
+    def get_model(self):
+        raise NotImplementedError("Subclasses must implement get_model method")
+
+    def compare(self, quantized_models, float_model, input_x=None, quantization_info=None):
```
Comment (irenaby) [inline tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py:87]: Consider implementing explicit compare in the concrete tests. IMO in tests simplicity and clarity is more important than a perfect OOP.

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 7]
```diff
@@ -41,8 +40,8 @@ def _extract_parameters_and_buffers(module: Union[torch.nn.Module, GraphModule],
         Dictionary containing module parameters and buffers by name.
     """
 
-    named_parameters = {name: to_numpy(parameter) for name, parameter in module.named_parameters()}
-    named_buffers = {name: to_numpy(buffer) for name, buffer in module.named_buffers()}
+    named_parameters = {name: parameter for name, parameter in module.named_parameters()}
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:43]: Can you explain why you removed numpy conversion? seems like something that might have effect on different modules, and is not directly related to reused layers

File: tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py [hunk 8]
```diff
@@ -61,7 +44,43 @@ def __init__(self, unit_test):
...
+
+        # Verify that the shared parameters have identical memory addresses
+        self.unit_test.assertEqual([p.data_ptr() for p in quant_model.conv1.parameters()],
+                                   [p.data_ptr() for p in quant_model.conv1_1.parameters()],
+                                   f"Shared parameters between reused layers should have identical memory addresses")
+
+        #########################################################################################
+
+        # Verify that 'conv1' is called twice (thus reused) and 'conv2' is called once
+        layer_calls = {}
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests/pytorch_tests/model_tests/feature_models/reuse_layer_net_test.py:100]: I would also suggest changing `fx_node_2_graph_node` to defaultdict.
Let's start improving all our code according to the recent principles that we presented, not just our new code :)

**Review by ofirgo — CHANGES_REQUESTED** [review]

**Review by ofirgo — APPROVED** [review]
